test/integration/service/service_test.go

```diff
@@ -406,3 +416,283 @@ func Test_ServiceClusterIPSelector(t *testing.T) {
+		if err != nil {
+			fmt.Fprintf(logsBuffer, "failed to list EndpointSlices for service %q: %v\n", svc.GetName(), err)
+			return false, nil
+		}
+		if slices == nil || len(slices.Items) == 0 {
+			fmt.Fprintf(logsBuffer, "no EndpointSlices returned for service %q\n", svc.GetName())
+			return false, nil
+		}
+		fmt.Fprintf(logsBuffer, "EndpointSlices=\n%v\n", format.Object(slices, 1 /* indent one level */))
+
+		for _, slice := range slices.Items {
+			for _, endpoint := range slice.Endpoints {
+				var ip string
+				if len(endpoint.Addresses) > 0 {
+					ip = endpoint.Addresses[0]
+				}
+				if endpoint.Hints != nil && len(endpoint.Hints.ForZones) != 0 {
+					fmt.Fprintf(logsBuffer, "endpoint with ip %v has hint %+v, want no hint\n", ip, endpoint.Hints)
+					return false, nil
+				}
+			}
+		}
+		return true, nil
+	}
+
+	err = wait.PollUntilContextTimeout(ctx, 1*time.Second, 10*time.Second, true, endpointSlicesHaveNoHints)
+	if err != nil {
+		t.Logf("logsBuffer=\n%v", logsBuffer)
+		t.Fatalf("Error waiting for EndpointSlices to have same zone hints: %v", err)
+	}
+	logsBuffer.Reset()
+
+	////////////////////////////////////////////////////////////////////////////
+	// Update the service by setting the `trafficDistribution: PreferLocal` field
+	//
+	// Assert that the respective EndpointSlices get the same-zone hints.
+	////////////////////////////////////////////////////////////////////////////
+
+	trafficDist := corev1.ServiceTrafficDistributionPreferClose
+	svc.Spec.TrafficDistribution = &trafficDist
+	_, err = client.CoreV1().Services(ns.Name).Update(ctx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to update test service with 'trafficDistribution: PreferLocal': %v", err)
+	}
+
+	endpointSlicesHaveSameZoneHints := func(ctx context.Context) (bool, error) {
+		slices, err := client.DiscoveryV1().EndpointSlices(ns.GetName()).List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1.LabelServiceName, svc.GetName())})
+		if err != nil {
+			fmt.Fprintf(logsBuffer, "failed to list EndpointSlices for service %q: %v\n", svc.GetName(), err)
+			return false, nil
+		}
+		if slices == nil || len(slices.Items) == 0 {
+			fmt.Fprintf(logsBuffer, "no EndpointSlices returned for service %q\n", svc.GetName())
+			return false, nil
+		}
+		fmt.Fprintf(logsBuffer, "EndpointSlices=\n%v\n", format.Object(slices, 1 /* indent one level */))
+
+		for _, slice := range slices.Items {
+			for _, endpoint := range slice.Endpoints {
+				var ip string
+				if len(endpoint.Addresses) > 0 {
+					ip = endpoint.Addresses[0]
+				}
+				var zone string
+				if endpoint.Zone != nil {
+					zone = *endpoint.Zone
+				}
+				if endpoint.Hints == nil || len(endpoint.Hints.ForZones) != 1 || endpoint.Hints.ForZones[0].Name != zone {
+					fmt.Fprintf(logsBuffer, "endpoint with ip %v does not have the correct hint, want hint for zone %q\n", ip, zone)
+					return false, nil
```

Comment:
error too, sme comment as before this does not seem something we should retry

Reply:
I guess this is also same as https://github.com/kubernetes/kubernetes/pull/123812#discussion_r1522146298

Reply:
yep, forget to update the comment sorry